vrclient: Add support for VROverlay (Fixes VR in ACC (805550))

mcoffin opened the pull request:
This set of patches adds the same texture translation that occurs in `VRCompositor::Submit` to `VROverlay::SetOverlayTexture` to fix the displaying of VR overlays.

There's a slight re-factor to move the responsibility of the texture translation to a separate function for the DXVK case, as it allows for good code re-use between VRCompositor and VROverlay.

There's also a single fix to `VRCompositor::Submit` to make it not segfault when `texture->handle` is `NULL`. Downstream implementations in SteamVR handle this case gracefully, so we should too, just passing through the NULL handle without segfaulting while trying to read the vptable.

This has been tested with [Assetto Corsa Competizione](https://store.steampowered.com/app/805550/Assetto_Corsa_Competizione/), and I can confirm that VR is working with that game in Proton with these patches on top of the latest 5.0.

This may fix VR usage for other applications that use OpenVR's `VROverlay`, but I haven't found any others yet. I'm thinking it may be common in UE4 games (the engine used by ACC)?

Sorry for the short and scattered PR message; I'm a bit scatter-brained today from lack of sleep, but I wanted to get this submitted before too long.

# Issue References

* [Assetto Corsa Competizione](https://store.steampowered.com/app/805550/Assetto_Corsa_Competizione/) - #1420

mcoffin committed:
vrclient/compositor: Fix segfault on NULL texture handles

`ivrcompositor_submit` will segfault when the handle of Texture_t is
NULL while it's checking to see what interfaces are implemented by it.

To avoid this, use the "invalid texture handle" code path that already
existed in this case.

This fixes a crash on startup for Assetto Corsa Competizione (805550)

mcoffin committed:
vrclient/overlay: Add support for IVROverlay for DXVK

mcoffin committed:
vrclient/overlay: Add support for setting Vulkan overlay textures

mcoffin commented:
cc: @jp7677

mcoffin commented on vrclient_x64/vrclient_x64/vrclient_main.c:
```diff
@@ -744,79 +769,171 @@ static EVRCompositorError ivrcompositor_submit_wined3d(
 }
 #endif
 
+static EVROverlayError ivroverlay_set_overlay_texture_vulkan(
```
Note: this seems simple, but I was unable to test as I don't know of a Vulkan VR program for windows that actually uses the `VROverlay` API from OpenVR.

farmboy0 commented:
Is there a build somewhere of proton which includes these patches?

mcoffin commented:
> Is there a build somewhere of proton which includes these patches?

@farmboy0 I haven't published binaries, as I never keep them up to date and don't want people using out-of-date stuff, but if you want a 1-time build I'd be happy to send one over via email, just toss me an email at [mcoffin13@gmail.com](mailto:mcoffin13@gmail.com).

doitsujin commented on vrclient_x64/vrclient_x64/vrclient_main.c:
```diff
@@ -891,6 +1008,49 @@ static EVRCompositorError ivrcompositor_submit_vulkan(
+					return ivroverlay_set_overlay_texture_dxvk(cpp_func, linux_side, overlayHandle, texture, version, user_data, dxvk_surface);
+				}
+			}
+#endif
```
We're implicitly falling through to the `TextureType_Vulkan` case, is that intended (and if so, why)?

mcoffin replied:
Looks like when I squashed the overlay_warn_out label got moved to the wrong spot. Functionally it’ll work because every path returns but you’re right that this wasn’t intended

mcoffin replied:
Resolved this with an amended commit

doitsujin commented on vrclient_x64/vrclient_x64/vrclient_main.c:
```diff
@@ -891,6 +1008,49 @@ static EVRCompositorError ivrcompositor_submit_vulkan(
+				goto overlay_warn_out;
+			}
+
+#ifdef VRCLIENT_HAVE_DXVK
```
Is there a wined3d code path? (do we want/need one at this point?)

mcoffin replied:
Oops, you’re totally right I missed that one. I’ll add it. Thanks!

doitsujin reviewed (commented):
See comments. I can't say much about the OpenVR side of things since I'm not too familiar with it, but the DXVK interaction looks like it should work fine.

aeikum commented:
Hey @mcoffin, saw you pushed a new commit yesterday. Do you think this is ready to merge?

mcoffin commented:
> Do you think this is ready to merge?

@aeikum I'm going to push one more quick fix right now to move the `overlay_warn_out` label to the right spot (no functional change), and test again quickly with ACC before pushing. We're still missing the VKD3D case right now, but since it's missing entirely before, that's not the end of the world. That said, I'll try to implement it, but if it doesn't work on the first shot, then I'll just add a `FIXME` message about VKD3D not being supported there yet.

I'll ping you when those 2 items are done. I'll get it in before EoD today. Sorry I've been neglecting this PR; work has been busy.

aeikum commented:
No worries at all, just checking whether the ball is in my court or not. I appreciate your work on this.

mcoffin committed:
vrclient/overlay: Add FIXME log for unsupported wined3d case

mcoffin commented:
@aeikum Ok, I did the above, but none of my test case games worked with wined3d, so I couldn't get that case working. I added a `FIXME` log for it, so that someone may implement it in the future. Since I'm locally testing off of `-next`, I rebased the PR off of the latest RC, and targeted the merge at `proton_5.0-next`.

This should be good to go now.

mcoffin commented:
The above force-push was just a conversion of some leftover tabs I accidentally added back into spaces :)

mcoffin commented:
@aeikum What do you all need from me at this point on this one? Sorry for the ping, just (like you) making sure the ball isn't in my court currently :)

aeikum commented:
Nothing at all! You mentioned a little bit ago that it's ready to go, so I've been working on merging it in. Definitely in my court, apologies I didn't make it clear.

frostworx commented:
For a quick test I tried the binaries found in the ACC issue with Summer Funland (780280) and it finally fixes [this bug ](https://github.com/ValveSoftware/Proton/issues/3292) (the game was only "playable" with launch parameter _-onethread)_.
Thank you very much for the patch @mcoffin!
_(I'm sure I'll find some more games, but this one was my favorite)_

aeikum commented:
Hi @mcoffin and @jp7677, I've included this patchset in the 5.13-2 Release Candidate that just went up. Can you give it a try to confirm it is working? I made some significant changes to your patchset, which you can see in this repository on the `proton_5.13-next` branch. Please see #4360 for instructions for running the Release Candidate.

frostworx commented:
Hi,
it wasn't requested, but I can confirm that Summer Funland (780280) works great with `proton-5.13-2-rc1`

jp7677 commented:
Hi @aeikum Thanks a lot for integrating this PR into 5.13-next. This branch works perfectly here in VR on Nvidia (well, with the exception of missing async reprojection). I could also no longer reproduce any game crashes, thus I think you have even fixed https://github.com/ValveSoftware/SteamVR-for-Linux/issues/397
Thanks a lot!

mimattr commented:
> Hi @aeikum Thanks a lot for integrating this PR into 5.13-next. This branch works perfectly here in VR on Nvidia (well, with the exception of missing async reprojection).

Agreed, many thanks to @mcoffin and @aeikum for this addition!

We just need NVIDIA to step-up with _"Async Reprojection"_ support so VR is actually viable on Linux with their hardware, is there a public ETA/roadmap for this feature?

aeikum commented:
This was merged in 5.13 as `ed6a644fc32f8606087d6fc0d698c5c07f6f154a^..a2ccc70f7b0ec522e02cd04e195cd7f574ff90a2`. Thanks!
